File: 0031-next-permutation/0031-next-permutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int n = nums.size();
        int breakpoint;
        // finding the breakpoint of decrement
        for(int i = n - 2; i >= 0; i--) {
            if(nums[i] < nums[i + 1]) {
                breakpoint = i;
                break;
            } 
        }
        /* breakpoint won't be found if the order is in decreasing ie last permutation.
        So we reverse the order to get the first permutation */
        if(breakpoint < 0) reverse(nums.begin(), nums.end());
        else {
            // finding the next greater number than the number found on breakpoint index
            for(int i = n - 1; i > breakpoint; i--) { // doing it reverse to get the increasing order
                if(nums[i] > nums[breakpoint]) {
                    swap(nums[i], nums[breakpoint]);
                    break;
                } 
            }
            // sorting the subarray after breakpoint so that the order gets increasing
            reverse(nums.begin() + breakpoint + 1, nums.end());
        }
    }
};

File: 0031-next-permutation/0031-next-permutation.java
class Solution {
    public void nextPermutation(int[] nums) {
        int n = nums.length;
        int breakpoint = -1;

        // 1. Find the breakpoint (first decreasing element from right)
        for (int i = n - 2; i >= 0; i--) {
            if (nums[i] < nums[i + 1]) {
                breakpoint = i;
                break;
            }
        }

        // 2. If no breakpoint, array is in descending order; just reverse it all
        if (breakpoint == -1) {
            reverse(nums, 0, n - 1);
            return;
        }

        // 3. Find the smallest number larger than nums[breakpoint] to its right
        for (int i = n - 1; i > breakpoint; i--) {
            if (nums[i] > nums[breakpoint]) {
                swap(nums, i, breakpoint); // Corrected swap arguments
                break;
            }
        }

        // 4. Reverse the portion after the breakpoint
        reverse(nums, breakpoint + 1, n - 1);
    }

    private void swap(int[] array, int i, int j) {
        int temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }

    private void reverse(int[] array, int start, int end) {
        while (start < end) {
            swap(array, start++, end--);
        }
    }
}